### PR TITLE
Gallery - fix Axe errors

### PR DESCRIPTION
**Pull request by alexanderPolosatov**

### CODE REVIEW

**Review by marker-dao — DISMISSED** [review]

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js [hunk 1]
```diff
@@ -78,6 +78,57 @@ QUnit.module('aria accessibility', () => {
         assert.equal($element.attr('aria-label'), 'gallery', 'widget should have aria-label to have difference from text list');
     });
 
+    QUnit.test('aria label for items', function(assert) {
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js]: [minor, test naming]
```suggestion
    QUnit.test('aria label for items should be set', function(assert) {
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js [hunk 2]
```diff
@@ -78,6 +78,57 @@ QUnit.module('aria accessibility', () => {
...
+        const $gallery = $('#gallerySimple').dxGallery({ items: [1] });
+
+        const $item = $gallery.find(`.${GALLERY_ITEM_CLASS}`);
+
+        assert.equal($item.attr('aria-label'), 'gallery item', 'item should have "aria-label"');
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js]: [code style]
```suggestion
        assert.strictEqual($item.attr('aria-label'), 'gallery item', 'item should have "aria-label"');
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js [hunk 3]
```diff
@@ -78,6 +78,57 @@ QUnit.module('aria accessibility', () => {
...
+        const $items = $gallery.find(`.${GALLERY_ITEM_CLASS}`);
+
+        const itemsWithId = Array.from($items).filter(item => {
+            return Boolean($(item).attr('id'));
+        });
+
+        if(windowUtils.hasWindow()) {
+            assert.strictEqual(itemsWithId.length, 1, 'id attribute should be exist only on one item');
+        } else {
+            assert.strictEqual(itemsWithId.length, 0, 'no rendered elements');
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js]: [base approach]
if this test is actual only not on a server we can move it to `gallery.tests.js`

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js [hunk 4]
```diff
@@ -78,6 +78,57 @@ QUnit.module('aria accessibility', () => {
...
+        }
+    });
+
+    QUnit.test('aria-activedescendant should have link to one element only when width was not set', function(assert) {
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js]: [unclear test]
but test above shows that activedescendant is set when width is set to `100%`.
I didn't catch how should it work.
If you wanna to test that activedescendant element points to the single elements in both cases when width is specified or nor specified, let's write a single test in `forEach`

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js [hunk 5]
```diff
@@ -385,15 +385,15 @@ QUnit.module('behavior', {
         assert.equal($loopItems.length, 2);
     });
 
-    QUnit.test('duplicate items should have role="presentation"', function(assert) {
+    QUnit.test('duplicate items should be hidden for readres', function(assert) {
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js]: [minor, test naming]
```suggestion
    QUnit.test('duplicate items should have aria-hidden=true', function(assert) {
```

File: packages/devextreme/js/ui/gallery.js [hunk 6]
```diff
@@ -146,7 +146,10 @@ const Gallery = CollectionWidget.inherit({
             */
 
 
-            _itemAttributes: { role: 'option' },
+            _itemAttributes: {
+                role: 'option',
+                'aria-label': 'gallery item'
```
Comment (ksercs) [inline packages/devextreme/js/ui/gallery.js]: [base approach]
We should use localization variables for labels

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js [hunk 7]
```diff
@@ -385,15 +385,15 @@ QUnit.module('behavior', {
...
             items: [0, 1, 2, 3],
             loop: true
         });
 
         const $loopItems = $gallery.find(`.${GALLERY_LOOP_ITEM_CLASS}`);
         $loopItems.each((index, item) => {
-            assert.strictEqual($(item).attr('role'), 'presentation');
+            assert.strictEqual($(item).attr('aria-hidden'), 'true');
         });
     });
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js:388]: [new, test coverage]
Let's add a testcafe test with Axe for base mods

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js (hunk 4 above)
Comment (alexanderPolosatov) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js]: fixed

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js (hunk 3 above)
Comment (alexanderPolosatov) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js]: fixed

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js (hunk 5 above)
Comment (alexanderPolosatov) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js]: fixed

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js (hunk 7 above)
Comment (alexanderPolosatov) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js:388]: fixed

File: packages/devextreme/js/ui/gallery.js (hunk 6 above)
Comment (alexanderPolosatov) [reply packages/devextreme/js/ui/gallery.js]: fixed

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js (hunk 2 above)
Comment (alexanderPolosatov) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js]: fixed

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js (hunk 1 above)
Comment (alexanderPolosatov) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.markup.tests.js]: fixed

**Review by marker-dao — DISMISSED** [review]

File: packages/devextreme/js/localization/messages/ar.json [hunk 8]
```diff
@@ -699,6 +699,8 @@
         "dxGantt-showResources": "Show Resources",
         "dxGantt-showDependencies": "Show Dependencies",
         "dxGantt-dialogStartDateValidation": "Start date must be after {0}",
-        "dxGantt-dialogEndDateValidation": "End date must be after {0}"
+        "dxGantt-dialogEndDateValidation": "End date must be after {0}",
+
+        "dxGalley-galleryItemName": "Gallery item"
```
Comment (ksercs) [inline packages/devextreme/js/localization/messages/ar.json]: xD
```suggestion
        "dxGallery-galleryItemName": "Gallery item"
```

File: packages/devextreme/js/ui/gallery.js [hunk 9]
```diff
@@ -273,10 +277,13 @@ const Gallery = CollectionWidget.inherit({
 
         this.callBase();
 
-        this.setAria({
-            'role': 'listbox',
+        const listBoxRoleShouldBe = this._itemsCount() > 0;
```
Comment (ksercs) [inline packages/devextreme/js/ui/gallery.js]:
```suggestion
        const useListBoxRole = this._itemsCount() > 0;
```

**Review by ksercs — CHANGES_REQUESTED** [review]
I'll continue review tomorrow

File: packages/devextreme/js/ui/gallery.js (hunk 9 above)
Comment (alexanderPolosatov) [reply packages/devextreme/js/ui/gallery.js]: fixed

File: packages/devextreme/js/localization/messages/ar.json (hunk 8 above)
Comment (alexanderPolosatov) [reply packages/devextreme/js/localization/messages/ar.json]: fixed

File: packages/devextreme/js/localization/messages/ar.json [hunk 10]
```diff
@@ -699,6 +699,8 @@
         "dxGantt-showResources": "Show Resources",
         "dxGantt-showDependencies": "Show Dependencies",
         "dxGantt-dialogStartDateValidation": "Start date must be after {0}",
-        "dxGantt-dialogEndDateValidation": "End date must be after {0}"
+        "dxGantt-dialogEndDateValidation": "End date must be after {0}",
+
+        "dxGallery-galleryItemName": "Gallery item"
```
Comment (ksercs) [inline packages/devextreme/js/localization/messages/ar.json]: Why we duplicate `gallery`?
Seems like there can not be other item except gallery item.
```suggestion
        "dxGallery-itemName": "Gallery item"
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js [hunk 11]
```diff
@@ -2762,3 +2751,55 @@ QUnit.module('gallery with paginated dataSource', {
...
+        const $items = $gallery.find(`.${GALLERY_ITEM_CLASS}`);
+
+        const itemsWithId = Array.from($items).filter(item => {
+            return Boolean($(item).attr('id'));
+        });
+
+        assert.strictEqual(itemsWithId.length, 1, 'id attribute should be exist only on one item');
+    });
+
+    QUnit.test('aria-activedescendant should have link to one element only after resize', function(assert) {
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js]:
```suggestion
    QUnit.test('aria-activedescendant should have link to the single element after resize', function(assert) {
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js [hunk 12]
```diff
@@ -2762,3 +2751,55 @@ QUnit.module('gallery with paginated dataSource', {
...
+
+        const $items = $gallery.find(`.${GALLERY_ITEM_CLASS}`);
+
+        const itemsWithId = Array.from($items).filter(item => {
+            return Boolean($(item).attr('id'));
+        });
+
+        assert.strictEqual(itemsWithId.length, 1, 'id attribute should be exist only on one item');
+    });
+
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js]:
```suggestion
```

File: packages/devextreme/testing/testcafe/tests/gallery/accessibility.ts [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+const testsSettings: {
+  testName: string;
+  gallerySettings: any;
+}[] = gallerySettings.map((settings) => ({
+  testName: getTestName(settings),
+  gallerySettings: settings,
+}));
+
+testsSettings.forEach((settings) => {
```
Comment (ksercs) [inline packages/devextreme/testing/testcafe/tests/gallery/accessibility.ts]: let's simplify all this stuff
```javascript
galleryConfigs.forEach(galleryConfig => {
    const testName = getTestName(settings);
    test(testName, async (t) => {
        await a11yCheck(t);
    }).before(async () => createWidget(
        'dxGallery',
         galleryConfig,
    ));
});
```

File: packages/devextreme/testing/testcafe/tests/gallery/accessibility.ts [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+  loop: true,
+}, {
+  showIndicator: false,
+}].reduce((acc: any, currentValue) => {
+  acc.push({
+    ...currentValue,
+    ...acc.at(-1),
+  });
+  return acc;
+}, []);
```
Comment (ksercs) [inline packages/devextreme/testing/testcafe/tests/gallery/accessibility.ts]: Let's write a usual brut force?
Now this return only 6 sets. But there are 2^5 different sets

File: packages/devextreme/js/localization/messages/ar.json (hunk 10 above)
Comment (alexanderPolosatov) [reply packages/devextreme/js/localization/messages/ar.json]: fixed

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js (hunk 11 above)
Comment (alexanderPolosatov) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js]: fixed

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js (hunk 12 above)
Comment (alexanderPolosatov) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/gallery.tests.js]: fixed

File: packages/devextreme/testing/testcafe/tests/gallery/accessibility.ts (hunk 14 above)
Comment (alexanderPolosatov) [reply packages/devextreme/testing/testcafe/tests/gallery/accessibility.ts]: reworked to brut force

File: packages/devextreme/testing/testcafe/tests/gallery/accessibility.ts (hunk 13 above)
Comment (alexanderPolosatov) [reply packages/devextreme/testing/testcafe/tests/gallery/accessibility.ts]: fixed

**Review by EugeniyKiyashko — APPROVED** [review]

**Review by ksercs — APPROVED** [review]

**Review by vladaskorohodova — APPROVED** [review]

**Pull request merged**